### PR TITLE
Introduce accessing organization and user resident organization attributes for the authenticated user.

### PR DESCRIPTION
**Pull request by sadilchamishka**
### Proposed changes in this pull request

A new attribute called `accessingOrganization`  & `userResidentOrganization` are introduced to the `AuthenticatedUser` object, which keeps the ID of the organization where the B2B user is authorized to access and the organization where the user's identity is managed. 

For federated users who login via the organization SSO authenticator (for B2B use cases), the `user_organization` claim is set in the user attributes by the authenticator. That value is taken as the authorizedOrganization as that organization where the B2B user intend to access. For other cases like B2C logins, `accessingOrganization` will be set to null. 

The `accessingOrganization` is reflected in the token table by the introduced column below.
A new column `AUTHORIZED_ORGANIZATION` is introduced to the `IDN_OAUTH2_ACCESS_TOKEN` table. That column implies the organization where the access token is bounded for access.

### Related Issues.
- https://github.com/wso2/product-is/issues/16365

Following figure shows how the authenticated user at the framework level propagated to the OAuth layer for persitening and wise versa. 
For B2B user logins, the authenticated user is treated as usual federated user. Two new attributes are added called accessingOrganization and userResidentOrganization.  Those information are stored in access token, and should be able to build the exact similar authenticated user object from the data stored in the token table.

- The consumer-key of the token table denotes the tenant-id which should represent in authenticed user object.
- The tenant ID of the token table is corresponds to the user's resident organization.

Screenshot 2023-10-14 at 23 30 36

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 1]
```diff
@@ -1420,4 +1427,29 @@ private void setLoggedInOrgIdInRequest(AuthenticatedIdPData authenticatedIdPData
             }
         }
     }
+
+    private void setUserOrganization(AuthenticatedUser authenticatedUser, AuthenticatorConfig authenticatorConfig) {
+
+        String userOrganization = null;
```
Comment (ShanChathusanda93) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:1433]:
```suggestion
        String userOrganization = StringUtils.EMPTY;
```
WDYT?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 2]
```diff
@@ -1420,4 +1427,29 @@ private void setLoggedInOrgIdInRequest(AuthenticatedIdPData authenticatedIdPData
...
+        if (authenticatedUser.getUserAttributes() != null && isLoggedInWithOrganizationLogin(authenticatorConfig)) {
+            for (Map.Entry<ClaimMapping, String> userAttributes : authenticatedUser.getUserAttributes().entrySet()) {
+                if (FrameworkConstants.USER_ORGANIZATION_CLAIM.equals(
+                        userAttributes.getKey().getLocalClaim().getClaimUri())) {
+                    userOrganization = userAttributes.getValue();
+                }
+            }
+        }
+
+        if (userOrganization == null) {
```
Comment (ShanChathusanda93) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:1443]:
```suggestion
        if (StringUtils.isEmpty(userOrganization)) {
```
WDYT?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 3]
```diff
@@ -1420,4 +1427,29 @@ private void setLoggedInOrgIdInRequest(AuthenticatedIdPData authenticatedIdPData
...
+            try {
+                userOrganization = FrameworkServiceDataHolder.getInstance().getOrganizationManager()
+                        .resolveOrganizationId(authenticatedUser.getTenantDomain());
+            } catch (OrganizationManagementClientException e) {
+                LOG.debug("Organization ID is not found for the given tenant domain");
+            } catch (OrganizationManagementException e) {
```
Comment (ShanChathusanda93) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:1449]: Why we need to catch these seperately?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java (hunk 1 above)
Comment (sadilchamishka) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: Yes, that should be better. But the implementation is bit changed after the latest commit including the proposed suggession. https://github.com/wso2/carbon-identity-framework/pull/4915/commits/d0ccd16eaae1d3ee40dcd94a133037ab8036a85b

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java (hunk 2 above)
Comment (sadilchamishka) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: Agree on the suggession. But the implementation was changed bit. Please review the method

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java (hunk 3 above)
Comment (sadilchamishka) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: The client exception is thrown when we request for organization ID for a tenant who doesn't have an org-id (b2b disabled tenants). Other exception is to handle the server error if occured.

Comment (sadilchamishka) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: The above implementation is now obsolete

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 4]
```diff
@@ -1421,4 +1431,36 @@ private void setLoggedInOrgIdInRequest(AuthenticatedIdPData authenticatedIdPData
...
+                if (FrameworkConstants.USER_ORGANIZATION_CLAIM.equals(
+                        userAttributes.getKey().getLocalClaim().getClaimUri())) {
+                    return userAttributes.getValue();
+                }
+            }
+        }
+        return StringUtils.EMPTY;
+    }
+
+    private String resolveTenantDomain(String organizationId) throws FrameworkException {
```
Comment (ShanChathusanda93) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:1457]: If we are going to add the method description to private methods, shall we add that to here as well?

Comment (sadilchamishka) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:1457]: yeah we can, but for private methods, the method description is not required unless its logic is hard to understand. But this method is simple. The above private method need more context info to understand.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java [hunk 5]
```diff
@@ -384,9 +384,12 @@ protected void concludeFlow(HttpServletRequest request, HttpServletResponse resp
             if (!sequenceConfig.getApplicationConfig().isSaaSApp()) {
                 String spTenantDomain = context.getTenantDomain();
                 String userTenantDomain = sequenceConfig.getAuthenticatedUser().getTenantDomain();
+                String authorizedOrganization = sequenceConfig.getAuthenticatedUser().getAuthorizedOrganization();
                 if (StringUtils.isNotEmpty(userTenantDomain)) {
+                    /*For non SaaS apps, user tenant domain and application tenant domain should be equal. But for
+                     organization bound tokens issued for B2B apps, the tenant domains should not be equal.*/
                     if (StringUtils.isNotEmpty(spTenantDomain) && !spTenantDomain.equals
-                            (userTenantDomain)) {
+                            (userTenantDomain) && authorizedOrganization == null) {
```
[the remark this answers is not in the document]
Comment (sadilchamishka) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4915/commits/56464839fc7926e105720a09f0e6ca488cd6c9c2

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java [hunk 6]
```diff
@@ -384,9 +384,12 @@ protected void concludeFlow(HttpServletRequest request, HttpServletResponse resp
             if (!sequenceConfig.getApplicationConfig().isSaaSApp()) {
                 String spTenantDomain = context.getTenantDomain();
                 String userTenantDomain = sequenceConfig.getAuthenticatedUser().getTenantDomain();
+                String authorizedOrganization = sequenceConfig.getAuthenticatedUser().getAuthorizedOrganization();
                 if (StringUtils.isNotEmpty(userTenantDomain)) {
+                    /*For non SaaS apps, user tenant domain and application tenant domain should be equal. But for
+                     organization bound tokens issued for B2B apps, the tenant domains should not be equal.*/
```
[the remark this answers is not in the document]
Comment (sadilchamishka) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4915/commits/56464839fc7926e105720a09f0e6ca488cd6c9c2

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 7]
```diff
@@ -1421,4 +1431,36 @@ private void setLoggedInOrgIdInRequest(AuthenticatedIdPData authenticatedIdPData
...
+
+        try {
+            return FrameworkServiceDataHolder.getInstance().getOrganizationManager()
+                    .resolveTenantDomain(organizationId);
+        } catch (OrganizationManagementException e) {
+            throw new FrameworkException(e.getMessage());
```
[the remark this answers is not in the document]
Comment (sadilchamishka) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4915/commits/56464839fc7926e105720a09f0e6ca488cd6c9c2

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 8]
```diff
@@ -1420,4 +1429,36 @@ private void setLoggedInOrgIdInRequest(AuthenticatedIdPData authenticatedIdPData
...
+
+        // For other users, the user reside organization is same as the corresponding organization of the tenant domain.
+        try {
+            OrganizationManager organizationManager = FrameworkServiceDataHolder.getInstance().getOrganizationManager();
+            if (organizationManager != null) {
+                return FrameworkServiceDataHolder.getInstance().getOrganizationManager()
+                        .resolveOrganizationId(authenticatedUser.getTenantDomain());
+            }
+        } catch (OrganizationManagementClientException e) {
+            LOG.debug("Organization ID is not found for the given tenant domain");
```
[the remark this answers is not in the document]
Comment (sadilchamishka) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: There can be tenants without organization ID. For those tenants resolving organization ID could cause issues. But this code block is now obsolete. Hence resolve this comment

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 9]
```diff
@@ -1421,4 +1431,36 @@ private void setLoggedInOrgIdInRequest(AuthenticatedIdPData authenticatedIdPData
...
+        // Check for user organization claim for the authenticated user via the organization SSO authenticator.
+        if (authenticatedUser.getUserAttributes() != null && isLoggedInWithOrganizationSSO(authenticatorConfig)) {
+            for (Map.Entry<ClaimMapping, String> userAttributes : authenticatedUser.getUserAttributes().entrySet()) {
+                if (FrameworkConstants.USER_ORGANIZATION_CLAIM.equals(
+                        userAttributes.getKey().getLocalClaim().getClaimUri())) {
+                    return userAttributes.getValue();
+                }
+            }
+        }
+        return StringUtils.EMPTY;
```
Comment (thanujalk) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:1454]: what is the scenario we get here?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java [hunk 10]
```diff
@@ -65,6 +65,7 @@ public class AuthenticatedUser extends User {
     private String authenticatedSubjectIdentifier;
     private String federatedIdPName;
     private boolean isFederatedUser;
+    private String authorizedOrganization;
```
Comment (sadilchamishka) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java:68]: accessingOrganization

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java [hunk 11]
```diff
@@ -65,6 +65,7 @@ public class AuthenticatedUser extends User {
     private String authenticatedSubjectIdentifier;
     private String federatedIdPName;
     private boolean isFederatedUser;
+    private String userOrganization;
```
[the remark this answers is not in the document]
Comment (sadilchamishka) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java]: This code is obsolete. There are two attributes `accessingOrganization` & `userResidentOrganization`

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java (hunk 9 above)
Comment (sadilchamishka) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4915/commits/7360e490007132ad9850fab163818ce8ba5a938c
This method is only executed for B2B apps. If B2C user logs from a B2B app, the organization ID of the tenant domain is returned.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java (hunk 10 above)
Comment (sadilchamishka) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/AuthenticatedUser.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4915/commits/7360e490007132ad9850fab163818ce8ba5a938c

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 12]
```diff
@@ -1421,4 +1429,26 @@ private void setLoggedInOrgIdInRequest(AuthenticatedIdPData authenticatedIdPData
...
+     *
+     * @param authenticatedUser   The authenticated user.
+     * @param authenticatorConfig The authenticator configurations.
+     * @return The organization where the user's identity is managed.
+     */
+    private String resolveUserResidentOrganization(AuthenticatedUser authenticatedUser,
+                                                   AuthenticatorConfig authenticatorConfig) throws FrameworkException {
+
+        // Check for user organization claim for the authenticated user via the organization login authenticator.
+        if (authenticatedUser.getUserAttributes() != null && isLoggedInWithOrganizationLogin(authenticatorConfig)) {
```
Comment (thanujalk) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java:1444]: You already checked whether user authenticated with org login
```suggestion
        if (authenticatedUser.getUserAttributes() != null) {
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java [hunk 13]
```diff
@@ -751,6 +751,14 @@ protected void doAuthentication(HttpServletRequest request, HttpServletResponse
                 return;
             }
 
+            // Set authorized organization and user resident organization for B2B user logins.
+            if (context.getSubject() != null && isLoggedInWithOrganizationLogin(authenticatorConfig)) {
+                String userResidentOrganization =
+                        resolveUserResidentOrganization(context.getSubject(), authenticatorConfig);
+                context.getSubject().setAccessingOrganization(userResidentOrganization);
```
[the remark this answers is not in the document]
Comment (sadilchamishka) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4915/commits/ae978f284202c078d3e24494b68339e7c653bae6

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java (hunk 12 above)
Comment (sadilchamishka) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/step/impl/DefaultStepHandler.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4915/commits/ae978f284202c078d3e24494b68339e7c653bae6

**Review by thanujalk — APPROVED** [review]

**Pull request merged**